### PR TITLE
refactor(compiler): enable `register` and `resolve` phases for local compilation

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This commit update the logic to enable `resolve` phase for local compilation. The `resolve` phase will be useful for local compilation in certain cases (will be used in followup PRs).

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts [hunk 1]
```diff
@@ -340,20 +340,6 @@ runInEachFileSystem(() => {
         }
       }
 
-      it('should not run resolve phase', () => {
```
Comment (pmvald) [inline packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts:343]: Can we keep this test to test that resolve phase actually runs (like by expecting `expect(handler.resolve).toHaveBeenCalled();`) ?

File: packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts [hunk 2]
```diff
@@ -340,20 +340,6 @@ runInEachFileSystem(() => {
...
-        spyOn(handler, 'resolve');
-        const {compiler, sourceFile} = setup(contents, [handler], CompilationMode.LOCAL);
-
-        compiler.analyzeSync(sourceFile);
-        compiler.resolve();
-
-        expect(handler.resolve).not.toHaveBeenCalled();
-      });
-
       it('should not register', () => {
```
Comment (pmvald) [inline packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts:343]: For my work (generating extra imports), we also need the register phase to run. Can we add it to this PR as well? Similar to the resolve phase, we can exit for all current implementations.

File: packages/compiler-cli/src/ngtsc/transform/src/compilation.ts [hunk 3]
```diff
@@ -465,6 +459,10 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
 
         trait = trait.toResolved(result.data ?? null, result.diagnostics ?? null);
 
+        if (this.compilationMode === CompilationMode.LOCAL) {
```
Comment (pmvald) [inline packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:462]: nit; Is this to skip the reexport analysis? It is more clear to me if we just wrap that logic in an `if (this.compilationMode !== CompilationMode.LOCAL)`, along with a comment why we don't need this in local compilation.

File: packages/compiler-cli/src/ngtsc/transform/src/compilation.ts [hunk 4]
```diff
@@ -484,7 +482,7 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
    * `ts.SourceFile`.
    */
   typeCheck(sf: ts.SourceFile, ctx: TypeCheckContext): void {
-    if (!this.fileToClasses.has(sf)) {
+    if (!this.fileToClasses.has(sf) || this.compilationMode === CompilationMode.LOCAL) {
```
Comment (pmvald) [inline packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:485]: Q: Do we ever call `typeCheck` in local compilation mode? I mean previously this check was not there and the typeCheck never got called (or it was being called ??)

File: packages/compiler-cli/src/ngtsc/transform/src/compilation.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/compiler-cli/src/ngtsc/transform/src/compilation.ts]: @pmvald I was trying to limit the amount of work performed in local compilation mode (since this part was never invoked before due to an early exit). Looking at the code, it looks "safe" for the local compilation, so I've dropped the check. Please let me know if any additional change is required.

File: packages/compiler-cli/src/ngtsc/transform/src/compilation.ts (hunk 4 above)
Comment (AndrewKushnir) [reply packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:480]: The `typeCheck` function was called, but the actual work was skipped because a class was not "resolved" in local compilation mode (since we didn't invoke "resolve" phase logic):
https://github.com/angular/angular/blob/main/packages/compiler-cli/src/ngtsc/transform/src/compilation.ts#L494-L495

File: packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts (hunk 2 above)
Comment (AndrewKushnir) [reply packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts]: Sure, I've updated the `register` phase logic in the same way.

File: packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/compiler-cli/src/ngtsc/transform/test/compilation_spec.ts:343]: Good point, I've reverted this part of the change and updated the code to verify that the `resolve` and `register` phases are invoked.

**Comment by AndrewKushnir** [comment]
@pmvald thanks for the feedback! I've updated PR to address all comments, please take another look when you get a chance. Thank you.

**Review by pmvald — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/transform/src/compilation.ts (hunk 3 above)
Comment (pmvald) [reply packages/compiler-cli/src/ngtsc/transform/src/compilation.ts]: It seems harmless to me. Is this logic related to generating those weird reexports aliases ('ɵng$google3...' stuff)? We plan to disable this whole aliasing for local compilation. I'll send a PR soon.

**Comment by AndrewKushnir** [comment]
[Presubmit](https://fusion2.corp.google.com/presubmit/598491372/OCL:598491372:BASE:598491299:1705293368450:2e472b21/targets).

File: packages/compiler-cli/src/ngtsc/transform/src/compilation.ts [hunk 5]
```diff
@@ -484,7 +477,7 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
...
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:480]: Minor style nit: since this early return is unrelated to the presence of the file in `this.fileToClasses`, it'd be more readable as a completely separate `if...return`.

**Review by alxhub — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/transform/src/compilation.ts (hunk 5 above)
Comment (AndrewKushnir) [reply packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:480]: Thanks Alex! I will update this in a followup PR, so we don't need to re-run the presubmit for this change.

**Comment by AndrewKushnir** [comment]
**Caretaker note**: the [presubmit](https://fusion2.corp.google.com/presubmit/598491372/OCL:598491372:BASE:598491299:1705293368450:2e472b21/targets) is "green", this PR is ready for merge.

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 08233b8374ab5300bd7a7028d6d28634d8a44820.
